src/client-side-metrics/operation-metrics-collector.ts

```diff
@@ -58,6 +72,35 @@ enum MetricsCollectorState {
   OPERATION_COMPLETE,
 }
 
+// This method swallows errors when metrics debugging is not enabled so
+// that errors don't bubble up to the user.
+function withMetricsDebug<T>(fn: () => T): T | undefined {
+  try {
+    return fn();
+  } catch (e) {
+    if (METRICS_DEBUG) {
+      throw e;
+    }
+  }
+  return;
+}
+
+// Checks that the state transition is valid and if not it throws a warning.
+function checkState<T>(
+  currentState: MetricsCollectorState,
+  allowedStates: MetricsCollectorState[],
+  fn: () => T
+): T | undefined {
+  if (allowedStates.includes(currentState)) {
+    return fn();
+  } else {
+    if (METRICS_DEBUG) {
+      throw Error('Invalid state transition');
+    }
```

Comment:
Since you're throwing errors now, this can be simplified quite a bit:
1. remove the `if (METRICS_DEBUG)` check here, and always throw an error if it's in the wrong state. `withMetricsDebug` will catch it and decide whether to raise. That keeps all the `METRICS_DEBUG` logic in one place
2. You now have the option to remove the whole `fn: () => T` callback part, and just return if it's in a valid state. Execution will just continue if the state is good, or end here if it's in a bad state. (But I know JS is more callback-centric than other languages, so maybe it's more idiomatic to keep it as-is. I find lots of nested callbacks hard to reason about, but maybe that's just me)

Reply:
> I find lots of nested callbacks hard to reason about, but maybe that's just me

I do too. I think this is a good suggestion.

> withMetricsDebug will catch it and decide whether to raise

This is true, but it does mean we should only use checkState when it is wrapped in withMetricsDebug.

Reply:
> This is true, but it does mean we should only use checkState when it is wrapped in withMetricsDebug.

Wasn't that the intention? That you wrap everything that could call in `withMetricsDebug`, and can then surpress exceptions?